### PR TITLE
create proc/func/trigger, create/drop database, transaction statements are not being tracked by the extensions which utilize processutility hook

### PR DESCRIPTION
**Pull request by kritika-0601**
Currently statements like create proc/func/trigger/database, drop database and transaction statements are not being tracked by the extensions which utilises the processutility hooks in Babelfish. Above issue happens because bbf_processutility doesn't call subsequent/previous processutility hook for above statements due to which these subsequent extension hooks are not being executed.
This commit resolves above issue by adding a custom processutility hook in engine to execute above Babelfish statements so that execution happens at last step of processutility hook chain which makes sure that subsequent extension hooks are executed.

Before:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> executes above statements -> execution completed/doesn't return to previous hook execution or doesn't execute subsequent hooks

After:
processUtility_hooks()-> ... -> bbf_processUtility_hook() -> ... -> processUtility_hooks()-> ... -> Standard_processUtility()-> executes above statements using custom hook -> returns to previous hook execution

Task: BABEL-4218

Since this PR includes calling the engine side function to extension side using hook, we do not require extra test scenarios to test it, it is already being covered in previous tests.
We have separate test PR for testing pg_stat_statements extension.

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/tcop/utility.c [hunk 1]
```diff
@@ -76,6 +76,7 @@
 
 /* Hook for plugins to get control in ProcessUtility() */
 ProcessUtility_hook_type ProcessUtility_hook = NULL;
+miscProcessUtility_hook_type miscProcessUtility_hook = NULL;
```
Comment (surendravishnoi) [inline src/backend/tcop/utility.c]: a better name would be bbfCustomProcessUtility_hook

File: src/include/tcop/utility.h [hunk 2]
```diff
@@ -16,6 +16,7 @@
 
 #include "tcop/cmdtag.h"
 #include "tcop/tcopprot.h"
+#include "parser/parse_node.h"
```
Comment (surendravishnoi) [inline src/include/tcop/utility.h]: why this header is required? specifically for which struct

File: src/backend/tcop/utility.c [hunk 3]
```diff
@@ -594,6 +596,15 @@ standard_ProcessUtility(PlannedStmt *pstmt,
 	pstate->p_sourcetext = queryString;
 	pstate->p_queryEnv = queryEnv;
 
+	if(miscProcessUtility_hook)
+		(*miscProcessUtility_hook)(pstate, pstmt, queryString, context, params, qc, &flag);
```
Comment (surendravishnoi) [inline src/backend/tcop/utility.c]: why not return true or false?

File: src/include/tcop/utility.h (hunk 2 above)
Comment (kritika-0601) [reply src/include/tcop/utility.h]: It is required for ParseState.

File: src/backend/tcop/utility.c [hunk 4]
```diff
@@ -559,6 +560,7 @@ standard_ProcessUtility(PlannedStmt *pstmt,
 	bool		isAtomicContext = (!(context == PROCESS_UTILITY_TOPLEVEL || context == PROCESS_UTILITY_QUERY_NONATOMIC) || IsTransactionBlock());
 	ParseState *pstate;
 	int			readonly_flags;
+	int flag = false;
```
Comment (surendravishnoi) [inline src/backend/tcop/utility.c]: not required

File: src/include/tcop/utility.h (hunk 2 above)
Comment (surendravishnoi) [reply src/include/tcop/utility.h]: please use forward declaration

File: src/include/tcop/utility.h [hunk 5]
```diff
@@ -108,5 +108,12 @@ CreateCommandName(Node *parsetree)
 extern LogStmtLevel GetCommandLogLevel(Node *parsetree);
 
 extern bool CommandIsReadOnly(PlannedStmt *pstmt);
+typedef bool (*bbfCustomProcessUtility_hook_type)(struct ParseState *pstate, PlannedStmt *pstmt, const char *queryString, ProcessUtilityContext context, 
+						  ParamListInfo params, QueryCompletion *qc);
+extern PGDLLIMPORT bbfCustomProcessUtility_hook_type bbfCustomProcessUtility_hook;
+
+void PLTsqlProcessTransaction(Node *parsetree,
```
Comment (surendravishnoi) [inline src/include/tcop/utility.h]: why have we added this declaration here?

**Review by surendravishnoi — APPROVED** [review]

**Review by HarshLunagariya — APPROVED** [review]

**Pull request merged**